### PR TITLE
feat(gatsby): support named splats in functions

### PR DESCRIPTION
**Pull request by KyleAMathews**
Switch from using path-to-regexp to @reach/router to do matches
so we're we use the same route matching lib as elsewhere + to add support for named splats (path-to-regexp only support `/foo/*` not `/foo/*bar`.

### CODE REVIEW

File: integration-tests/functions/__tests__/__snapshots__/functions-dev.js.snap [hunk 1]
```diff
@@ -1,5 +1,25 @@
...
+Object {
+  "super": "additional",
+  "userId": "23",
+}
+`;
+
+exports[`develop dynamic routes unnamed wildcard routes 1`] = `
+Object {
+  "*": "super",
+  "0": "super",
```
Comment (KyleAMathews) [inline integration-tests/functions/__tests__/__snapshots__/functions-dev.js.snap:19]: send the path-to-regexp param style route so we avoid a breaking change until v5.

**Comment by LekoArts** [comment]
Can you please also update any relevant documentation for this? I think we somewhere note that we don't support this and/or the syntax needs change.

**Comment by KyleAMathews** [comment]
We never did support dynamic routing for functions in cloud so we [removed the docs for launch](https://github.com/gatsbyjs/gatsby/pull/31809/files#diff-6739eabfb34bc269c34a222cc66d3876732d744c6b6335e31fc0a63750474cbdL16) — @dmccraw is working on adding that which actually is what prompted noticing that we were missing support for named splats in the framework. When we add back the docs, let's make sure to add docs for this too.

**Comment by njbmartin** [comment]
Great to see this being supported! I tried using dynamic routing in my entry for the FuncJam.
https://github.com/njbmartin/gatsby-funcjam-21#dynamic-api-routing

TLDR;
> I misread the documentation (https://www.gatsbyjs.com/docs/reference/functions/routing/) ("Function routing shares the same syntax as page routing."). Oddly enough, it worked locally, even when serving a build, but doesn't work in Gatsby Cloud... sad face emoji

**Comment by wardpeet** [comment]
Seems like a valid error in the integration tests. I would also suggest moving to inlineSnapshot for the one small snapshots, it will make the tests a little bit more clear

**Comment by KyleAMathews** [comment]
Turns out I'd just missed committing my new function file 🤦
